OrbitBase/ThreadPool.cpp

```diff
@@ -103,38 +140,40 @@ std::unique_ptr<Action> ThreadPoolImpl::TakeAction() {
   std::unique_ptr<Action> action = std::move(scheduled_actions_.front());
   scheduled_actions_.pop_front();
 
-  CHECK(idle_threads_ > 0);  // Sanity check
-  // Note: do not move this outside of this method. This has to be done in
-  // the same lock scope as taking an action from the queue to avoid a situation
-  // where the check in Schedule() does not create a worker thread because if it
-  // is called between taking action from a queue and reducing idle_thread_
-  // counter the idle_threads_ < scheduled_actions_.size() check will
-  // be false and the action will end up being stuck in the queue for some
-  // time.
-  --idle_threads_;
-
   return action;
 }
 
 void ThreadPoolImpl::WorkerFunction() {
   while (true) {
+    absl::MutexLock lock(&mutex_);
     std::unique_ptr<Action> action = TakeAction();
 
     if (!action) {
+      // Move this thread from the worker_threads_ to finished_threads_.
+      std::thread::id thread_id = std::this_thread::get_id();
+      auto it = worker_threads_.find(thread_id);
+      CHECK(it != worker_threads_.end());
+      finished_threads_.push_back(std::move(it->second));
+      worker_threads_.erase(it);
+
       break;
```

Comment:
Should we decrease idle_threads_ when marking a thread as finished?

Reply:
yep, that's a bug - I should write a test for it.

Reply:
added test and fixed the bug.